File: Sources/Sideros/Models/Qwen2.swift
import Foundation
import MLX

public struct Qwen2Activations {
    public let embeddings: MLXArray
    public let blocks: [MLXArray]
    public let norm: MLXArray
    public let logits: MLXArray
}

public struct Qwen2: CausalLM {
    private let config: Qwen2Config
    private let parameters: [String: MLXArray]
    private let quantization: Quantization?
    private let skinny: Bool
    /// One compiled closure per layer: the trace fuses silu(gate) · up into one kernel.
    /// Decode re-traces only when the input shape changes — once for the prefill shape,
    /// once for the single-token shape.
    let mlps: [@Sendable (MLXArray) -> MLXArray]

    public init(
        config: Qwen2Config, parameters: [String: MLXArray], quantization: Quantization? = nil,
        skinny: Bool = false
    ) {
        self.config = config
        self.quantization = quantization
        self.skinny = skinny

        // One fused projection per layer instead of three: q, k and v are concatenated
        // along the output axis, which is row-aligned for every representation — dense
        // (out, in), packed u32, scales/biases per row group, and the projection bias.
        var parameters = parameters
        for layer in 0..<config.numHiddenLayers {
            let p = "model.layers.\(layer).self_attn."
            for suffix in ["weight", "scales", "biases", "bias"] {
                let parts = ["q_proj.", "k_proj.", "v_proj."].compactMap {
                    parameters[p + $0 + suffix]
                }
                guard parts.count == 3 else { continue }
                parameters[p + "qkv_proj." + suffix] = concatenated(parts, axis: 0)
            }
        }
        self.parameters = parameters

        mlps = (0..<config.numHiddenLayers).map { layer in
            let p = "model.layers.\(layer).mlp."
            let gate = Linear(
                p + "gate_proj", parameters: parameters, quantization: quantization, skinny: skinny)
            let up = Linear(
                p + "up_proj", parameters: parameters, quantization: quantization, skinny: skinny)
            let down = Linear(
                p + "down_proj", parameters: parameters, quantization: quantization, skinny: skinny)
            // silu spelled out: MLXNN's is itself a compiled function with its own lock,
            // and entering it while this trace holds the global eval lock inverts the
            // order against any thread calling it directly — an ABBA deadlock.
            return compile { x in
                let gated = gate(x)
                return down(gated * sigmoid(gated) * up(x))
            }
        }
    }

    /// Total by construction: loadQwen2Parameters rejects any checkpoint missing a name.
    private subscript(name: String) -> MLXArray { parameters[name]! }

    private func linear(_ name: String) -> Linear {
        Linear(name, parameters: parameters, quantization: quantization, skinny: skinny)
    }

    /// No positional embedding table: position enters only through RoPE.
    public func embed(_ tokens: MLXArray, offset: Int = 0) -> MLXArray {
        let table = Embedding("model.embed_tokens", parameters: parameters, quantization: quantization)
        return table(tokens)
    }

    public func makeCache() -> [KVCache] { (0..<config.numHiddenLayers).map { _ in KVCache() } }

    public func logits(_ embeddings: MLXArray, cache: [KVCache])
        -> (logits: MLXArray, cache: [KVCache])
    {
        let decoded = decode(embeddings, cache: cache)
        return (decoded.logits, decoded.cache)
    }

    public func callAsFunction(_ ids: [Int32]) -> Qwen2Activations {
        let embeddings = embed(ids)
        let decoded = decode(embeddings, cache: makeCache())
        return Qwen2Activations(
            embeddings: embeddings,
            blocks: decoded.blocks,
            norm: decoded.norm,
            logits: decoded.logits
        )
    }

    private func decode(_ embeddings: MLXArray, cache: [KVCache])
        -> (blocks: [MLXArray], norm: MLXArray, logits: MLXArray, cache: [KVCache])
    {
        var cache = cache
        var x = embeddings
        var blocks: [MLXArray] = []
        for layer in 0..<config.numHiddenLayers {
            (x, cache[layer]) = block(x, layer: layer, cache: cache[layer])
            blocks.append(x)
        }

        let normed = rmsNorm(x, prefix: "model.norm")
        return (blocks, normed, linear("model.embed_tokens")(normed), cache)  // tied lm_head
    }

    func block(_ x: MLXArray, layer: Int, cache: KVCache) -> (MLXArray, KVCache) {
        let prefix = "model.layers.\(layer)."
        let (context, grown) = attention(
            rmsNorm(x, prefix: prefix + "input_layernorm"), prefix: prefix, cache: cache)
        let attended = x + context
        return (
            attended
                + mlps[layer](rmsNorm(attended, prefix: prefix + "post_attention_layernorm")),
            grown
        )
    }

    func attention(_ x: MLXArray, prefix: String, cache: KVCache) -> (MLXArray, KVCache) {
        let length = x.dim(1)
        let offset = cache.offset
        let p = prefix + "self_attn."

        let queryWidth = config.numAttentionHeads * config.headDim
        let keyValueWidth = config.numKeyValueHeads * config.headDim
        let fused = linear(p + "qkv_proj")(x)
        let parts = split(fused, indices: [queryWidth, queryWidth + keyValueWidth], axis: -1)

        func heads(_ part: MLXArray, _ count: Int) -> MLXArray {
            part.reshaped(1, length, count, config.headDim).transposed(0, 2, 1, 3)
        }

        let queries = rope(heads(parts[0], config.numAttentionHeads), offset: offset)
        let (keys, values, grown) = cache.appending(
            keys: rope(heads(parts[1], config.numKeyValueHeads), offset: offset),
            values: heads(parts[2], config.numKeyValueHeads))

        // GQA: the kernel broadcasts the 2 kv heads across the 14 query heads itself.
        // `.causal` skips the masked blocks outright, rather than adding -3.4e38 and
        // letting softmax cancel it — the two differ in the last bits. It aligns the
        // queries to the end of the keys, so it is already right for a cached step.
        // A single query is the last position and may attend to everything: no mask.
        let context = MLXFast.scaledDotProductAttention(
            queries: queries, keys: keys, values: values,
            scale: 1 / sqrt(Float(config.headDim)), mask: length == 1 ? .none : .causal)

        // o_proj is the one projection without a bias.
        let projected = linear(p + "o_proj")(
            context.transposed(0, 2, 1, 3).reshaped(1, length, config.hiddenSize))
        return (projected, grown)
    }

    func rmsNorm(_ x: MLXArray, prefix: String) -> MLXArray {
        MLXFast.rmsNorm(x, weight: self[prefix + ".weight"], eps: config.rmsNormEps)
    }

    /// `traditional: false` is the interleaving HF calls `rotate_half`: the vector is
    /// split in two halves rather than into adjacent pairs.
    func rope(_ x: MLXArray, offset: Int = 0) -> MLXArray {
        MLXFast.RoPE(
            x, dimensions: config.headDim, traditional: false, base: config.ropeTheta,
            scale: 1, offset: offset)
    }
}

File: packages/sideros/src/sideros/models/qwen2.py
"""Qwen2 dense: Qwen3 minus the per-head q/k RMSNorm and minus explicit head_dim,
plus a bias on q/k/v.

Authoritative semantics: transformers' modeling_qwen2.py. `head_dim` is not in the
config — it is `hidden_size // num_attention_heads` (0.5B: 896 // 14 = 64) — and the
three attention projections carry a bias (`o_proj` does not; the MLP does not). Qwen2
is the only architecture in the house with it.

The qkv load-time fusion is a concatenation on the output axis, and rows are the
fusion axis in every representation the checkpoint uses: dense weight, packed u32
plus scales/biases, and the projection bias vector (whose single axis *is* the output
axis). One fusion rule covers all four.
"""

import json
import math
from dataclasses import dataclass
from pathlib import Path
from typing import NamedTuple, TypedDict

import mlx.core as mx
import mlx.nn as nn

from sideros.bpe import ByteLevelBPE
from sideros.chat import chat_capabilities
from sideros.checkpoint import (
    checkpoint,
    concat_gate_up,
    fuse_qkv,
    load_shards,
    prepare_weights,
    stop_tokens,
)
from sideros.core.cache import KVCache
from sideros.core.layers import SwiGLU, split_qkv
from sideros.language import LanguageModel, TextLanguageModel
from sideros.model import CompositeModel, ModelInput


@dataclass(frozen=True)
class Qwen2Config:
    hidden_size: int
    num_hidden_layers: int
    num_attention_heads: int
    num_key_value_heads: int
    vocab_size: int
    rms_norm_eps: float
    rope_theta: float
    tie_word_embeddings: bool
    intermediate_size: int
    eos_token_id: tuple[int, ...]

    @property
    def head_dim(self) -> int:
        return self.hidden_size // self.num_attention_heads


class Qwen2Attention(nn.Module):
    def __init__(self, config: Qwen2Config) -> None:
        super().__init__()
        self.heads = config.num_attention_heads
        self.kv_heads = config.num_key_value_heads
        self.head_dim = config.head_dim
        self.rope_theta = config.rope_theta
        hidden = config.hidden_size
        key_values = self.kv_heads * self.head_dim
        self.qkv_proj = nn.Linear(hidden, hidden + 2 * key_values, bias=True)
        self.o_proj = nn.Linear(hidden, hidden, bias=False)

    def split_heads(self, x: mx.array) -> tuple[mx.array, mx.array, mx.array]:
        """qkv split back into per-head [1, heads, length, head_dim], unrotated — the
        boundary transformers' q_proj/k_proj hooks expose, modulo the head reshape."""
        return split_qkv(
            self.qkv_proj(x),
            heads=self.heads,
            kv_heads=self.kv_heads,
            head_dim=self.head_dim,
        )

    def rope(self, x: mx.array, offset: int) -> mx.array:
        return mx.fast.rope(
            x, self.head_dim, traditional=False, base=self.rope_theta, scale=1.0, offset=offset
        )

    def __call__(self, x: mx.array, cache: KVCache) -> mx.array:
        length = x.shape[1]
        offset = cache.offset
        q, k, v = self.split_heads(x)
        keys, values = cache.update_and_fetch(self.rope(k, offset), v)
        attended = mx.fast.scaled_dot_product_attention(
            self.rope(q, offset), keys, values,
            scale=1 / math.sqrt(self.head_dim),
            mask=None if length == 1 else "causal",
        )
        return self.o_proj(
            attended.transpose(0, 2, 1, 3).reshape(1, length, self.heads * self.head_dim)
        )


class Qwen2Block(nn.Module):
    def __init__(self, config: Qwen2Config) -> None:
        super().__init__()
        self.self_attn = Qwen2Attention(config)
        self.mlp = SwiGLU(config.hidden_size, config.intermediate_size)
        self.input_layernorm = nn.RMSNorm(config.hidden_size, eps=config.rms_norm_eps)
        self.post_attention_layernorm = nn.RMSNorm(config.hidden_size, eps=config.rms_norm_eps)

    def __call__(self, x: mx.array, cache: KVCache) -> mx.array:
        attended = x + self.self_attn(self.input_layernorm(x), cache)
        return attended + self.mlp(self.post_attention_layernorm(attended))


class Qwen2Trunk(nn.Module):
    def __init__(self, config: Qwen2Config) -> None:
        super().__init__()
        self.embed_tokens = nn.Embedding(config.vocab_size, config.hidden_size)
        self.layers = [Qwen2Block(config) for _ in range(config.num_hidden_layers)]
        self.norm = nn.RMSNorm(config.hidden_size, eps=config.rms_norm_eps)


class Qwen2Activations(NamedTuple):
    embeddings: mx.array
    blocks: list[mx.array]
    norm: mx.array
    logits: mx.array


class Qwen2(nn.Module):
    def __init__(self, config: Qwen2Config) -> None:
        super().__init__()
        self.config = config
        self.model = Qwen2Trunk(config)
        if not config.tie_word_embeddings:
            self.lm_head = nn.Linear(config.hidden_size, config.vocab_size, bias=False)

    def make_cache(self) -> list[KVCache]:
        return [KVCache() for _ in self.model.layers]

    def head(self, normed: mx.array) -> mx.array:
        if self.config.tie_word_embeddings:
            return self.model.embed_tokens.as_linear(normed)
        return self.lm_head(normed)

    def activations(self, ids: mx.array, cache: list[KVCache] | None = None) -> Qwen2Activations:
        cache = cache if cache is not None else self.make_cache()
        x = self.model.embed_tokens(ids)
        embeddings = x
        blocks: list[mx.array] = []
        for block, layer_cache in zip(self.model.layers, cache, strict=True):
            x = block(x, layer_cache)
            blocks.append(x)
        normed = self.model.norm(x)
        return Qwen2Activations(embeddings, blocks, normed, self.head(normed))

    def __call__(self, ids: mx.array, cache: list[KVCache] | None = None) -> mx.array:
        return self.activations(ids, cache).logits


class _Json(TypedDict):
    model_type: str
    hidden_size: int
    num_hidden_layers: int
    num_attention_heads: int
    num_key_value_heads: int
    vocab_size: int
    rms_norm_eps: float
    rope_theta: float
    tie_word_embeddings: bool
    intermediate_size: int
    eos_token_id: int | list[int]


def _config(path: Path) -> Qwen2Config:
    raw: _Json = json.loads(path.read_text())
    if raw["model_type"] != "qwen2":
        raise ValueError(f"expected model_type qwen2, got {raw['model_type']!r}")
    eos = raw["eos_token_id"]
    return Qwen2Config(
        hidden_size=raw["hidden_size"],
        num_hidden_layers=raw["num_hidden_layers"],
        num_attention_heads=raw["num_attention_heads"],
        num_key_value_heads=raw["num_key_value_heads"],
        vocab_size=raw["vocab_size"],
        rms_norm_eps=raw["rms_norm_eps"],
        rope_theta=raw["rope_theta"],
        tie_word_embeddings=raw["tie_word_embeddings"],
        intermediate_size=raw["intermediate_size"],
        eos_token_id=tuple(eos) if isinstance(eos, list) else (eos,),
    )


def _weights(directory: Path, config: Qwen2Config, dtype: mx.Dtype | None) -> dict[str, mx.array]:
    layers = config.num_hidden_layers
    return prepare_weights(
        config,
        load_shards(directory),
        [
            lambda weights: fuse_qkv(weights, layers),
            lambda weights: concat_gate_up(weights, layers),
        ],
        dtype,
    )


def _composite(directory: Path, model: Qwen2) -> LanguageModel[ModelInput]:
    return CompositeModel(
        TextLanguageModel(
            model,
            ByteLevelBPE.from_file(directory / "tokenizer.json"),
            stop=stop_tokens(directory, model.config.eos_token_id),
        ),
        chat_capabilities(directory),
    )


CHECKPOINT = checkpoint((
        "config.json",
        "model*.safetensors",
        "tokenizer.json",
        "tokenizer_config.json",
        "chat_template.jinja",
    ), _config, Qwen2, _weights, _composite)
